File: src/main/java/com/nerdearla/workshop/paymentMethod/validator/PaymentMethodValidator.java
package com.nerdearla.workshop.paymentMethod.validator;

import com.nerdearla.workshop.payment.service.model.PaymentMethodData;
import com.nerdearla.workshop.paymentMethod.BuyerPaymentMethod;
import com.nerdearla.workshop.paymentMethod.error.DisabledPaymentMethodError;
import com.nerdearla.workshop.paymentMethod.error.InvalidSecurityCodeError;
import com.nerdearla.workshop.paymentMethod.error.InvalidTokenError;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Component;

@Component
public class PaymentMethodValidator {

    private static final Logger LOGGER = LoggerFactory.getLogger(PaymentMethodValidator.class);

    public void validate(BuyerPaymentMethod buyerPaymentMethod, PaymentMethodData paymentMethodData) {
        validateIsEnabled(buyerPaymentMethod);
        validateSecurityCode(buyerPaymentMethod.getSecurityCode(), paymentMethodData.getSecurityCode());
        validateToken(buyerPaymentMethod.getToken(), paymentMethodData.getToken());
    }

    private void validateIsEnabled(BuyerPaymentMethod buyerPaymentMethod) {
        if (Boolean.FALSE.equals(buyerPaymentMethod.getEnabled())) {
            LOGGER.error("Payment method {} is not enabled", buyerPaymentMethod.getId());
            throw new DisabledPaymentMethodError();
        }
        LOGGER.info("Payment method {} is enabled", buyerPaymentMethod.getId());
    }

    private void validateSecurityCode(String expected, String actual) {
        if (!expected.equals(actual)) {
            LOGGER.error("securityCode {} doesnt match expectation", actual);
            throw new InvalidSecurityCodeError();
        }
        LOGGER.info("securityCode validated successfully");
    }

    private void validateToken(String expected, String actual) {
        if (!expected.equals(actual)) {
            LOGGER.error("token {} doesnt match expectation", actual);
            throw new InvalidTokenError();
        }
        LOGGER.info("token validated successfully");
    }
}

File: src/main/java/com/nerdearla/workshop/paymentMethod/validator/PaymentMethodValidator.kt
package com.nerdearla.workshop.paymentMethod.validator

import com.nerdearla.workshop.payment.service.model.PaymentMethodData
import com.nerdearla.workshop.paymentMethod.BuyerPaymentMethod
import com.nerdearla.workshop.paymentMethod.error.DisabledPaymentMethodError
import com.nerdearla.workshop.paymentMethod.error.InvalidSecurityCodeError
import com.nerdearla.workshop.paymentMethod.error.InvalidTokenError
import com.nerdearla.workshop.shared.utils.CompanionLogger
import org.springframework.stereotype.Component

@Component
class PaymentMethodValidator {

    fun validate(buyerPaymentMethod: BuyerPaymentMethod, paymentMethodData: PaymentMethodData) {
        validateIsEnable(buyerPaymentMethod)
        validateSecurityCode(buyerPaymentMethod.securityCode, paymentMethodData.securityCode)
        validateToken(buyerPaymentMethod.token, paymentMethodData.token)
    }

    private fun validateIsEnable(buyerPaymentMethod: BuyerPaymentMethod) {
        if (buyerPaymentMethod.enabled.not()) {
            log.error("payment method {} is disabled", buyerPaymentMethod.id)
            throw DisabledPaymentMethodError()
        }

        log.info("payment method {} is enabled", buyerPaymentMethod.id)
    }

    private fun validateSecurityCode(expected: String, actual: String) {
        if (expected != actual) {
            log.error("securityCode {} doesnt match expectation", actual)
            throw InvalidSecurityCodeError()
        }

        log { info("securtyCode validated successfully") }
    }

    private fun validateToken(expected: String, actual: String) {
        if (expected != actual) {
            log.error("token {} doesnt match expectation", actual)
            throw InvalidTokenError()
        }

        log { info("token validated successfully") }
    }

    companion object : CompanionLogger()
}
